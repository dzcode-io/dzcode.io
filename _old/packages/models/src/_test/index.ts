import { OptionalPropertiesOf, RequiredPropertiesOf } from "@dzcode.io/utils/dist/ts";
import { ClassConstructor, plainToClass } from "class-transformer";
import { validateSync } from "class-validator";
import { BaseEntity } from "src/_base";

export type Cases<T extends object> = [
  [string, RequiredPropertiesOf<T>, false],
  [string, Required<T>, false],
  [string, Record<string, never>, boolean],
];

export const runDTOTestCases = <T extends BaseEntity>(
  entity: ClassConstructor<T>,
  requiredFields: RequiredPropertiesOf<T>,
  optionalFields: OptionalPropertiesOf<T>,
): void => {
  const hasNoRequiredFields = Object.keys(requiredFields).length === 0;
  const cases: Cases<T> = [
    ["should match snapshot when providing required fields only", requiredFields, false],
    [
      "should match snapshot when providing all fields",
      { ...requiredFields, ...optionalFields } as Required<T>,
      false,
    ],
    [
      hasNoRequiredFields
        ? "should match snapshot when providing an empty object"
        : "should show an error that matches snapshot when passing empty object",
      {},
      !hasNoRequiredFields,
    ],
  ];

  it.each(cases)("%s", (name, input, expectErrors) => {
    const output = plainToClass(entity, input);
    const errors = validateSync(output);

    expect(output).toMatchSnapshot("output");
    expect(errors).toMatchSnapshot("errors");
    if (expectErrors) {
      expect(errors.length).not.toBeLessThanOrEqual(0);
    } else {
      expect(errors.length).toBe(0);
    }
  });
};
