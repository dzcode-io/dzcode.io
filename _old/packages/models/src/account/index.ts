import { Type } from "class-transformer";
import { IsString, IsUrl, ValidateNested } from "class-validator";
import { BaseEntity, Model } from "src/_base";
import { RepositoryReferenceEntity } from "src/repository-reference";

export class AccountEntity extends BaseEntity {
  @IsString()
  id!: string;

  @IsString()
  username!: string;

  @IsString()
  name!: string;

  // eslint-disable-next-line camelcase
  @IsUrl({ require_protocol: false, require_host: false })
  profileUrl!: string;

  @IsUrl()
  avatarUrl!: string;

  @ValidateNested({ each: true })
  @Type(() => RepositoryReferenceEntity)
  repositories?: Model<RepositoryReferenceEntity>[];
}
