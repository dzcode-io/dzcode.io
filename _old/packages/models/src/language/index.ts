import { IsIn } from "class-validator";
import { BaseEntity } from "src/_base";

export const allLanguages = [
  { code: "en", shortLabel: "EN", label: "English", direction: "ltr" },
  { code: "ar", shortLabel: "ع", label: "العربية", direction: "rtl" },
] as const;

export class LanguageEntity extends BaseEntity {
  @IsIn(allLanguages.map(({ code }) => code))
  code!: typeof allLanguages[number]["code"];

  @IsIn(allLanguages.map(({ shortLabel }) => shortLabel))
  shortLabel!: typeof allLanguages[number]["shortLabel"];

  @IsIn(allLanguages.map(({ label }) => label))
  label!: typeof allLanguages[number]["label"];

  @IsIn(allLanguages.map(({ direction }) => direction))
  direction!: typeof allLanguages[number]["direction"];
}
