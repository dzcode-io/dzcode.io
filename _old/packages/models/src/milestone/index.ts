import { IsDateString, IsNumber, IsOptional, IsString, IsUrl } from "class-validator";
import { BaseEntity } from "src/_base";

export class MilestoneEntity extends BaseEntity {
  @IsString()
  id!: string;

  @IsString()
  title!: string;

  @IsString()
  description!: string;

  @IsUrl()
  url!: string;

  @IsNumber()
  openIssuesCount!: number;

  @IsNumber()
  closedIssuesCount!: number;

  @IsString()
  status!: "open" | "closed" | "in-progress";

  @IsDateString()
  createdAt!: string;

  @IsDateString()
  @IsOptional()
  dueAt?: string;

  @IsDateString()
  @IsOptional()
  closedAt?: string;
}
