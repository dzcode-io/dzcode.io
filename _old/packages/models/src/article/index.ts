import { Type } from "class-transformer";
import { IsString, ValidateNested } from "class-validator";
import { BaseEntity, Model } from "src/_base";
import { AccountEntity } from "src/account";

export class ArticleInfoEntity extends BaseEntity {
  @IsString()
  slug!: string;

  @IsString()
  title!: string;
}

export class ArticleEntity extends ArticleInfoEntity {
  @IsString()
  image!: string;

  @IsString()
  description!: string;

  @IsString()
  content!: string;

  @ValidateNested({ each: true })
  @Type(() => AccountEntity)
  authors!: Model<AccountEntity>[];

  @ValidateNested({ each: true })
  @Type(() => AccountEntity)
  contributors!: Model<AccountEntity>[];
}
