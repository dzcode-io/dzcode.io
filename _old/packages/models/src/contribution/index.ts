import { Type } from "class-transformer";
import { IsDateString, IsNumber, IsString, IsUrl, ValidateNested } from "class-validator";
import { BaseEntity, Model } from "src/_base";
import { AccountEntity } from "src/account";
import { ProjectReferenceEntity } from "src/project-reference";

export class ContributionEntity extends BaseEntity {
  @IsString()
  id!: string;

  @IsString()
  title!: string;

  @ValidateNested()
  @Type(() => ProjectReferenceEntity)
  project!: Model<ProjectReferenceEntity>;

  @ValidateNested()
  @Type(() => AccountEntity)
  createdBy!: Model<AccountEntity>;

  @IsString()
  type!: "issue" | "pullRequest";

  @IsUrl()
  url!: string;

  @IsString({ each: true })
  languages!: string[];

  @IsString({ each: true })
  labels!: string[];

  @IsDateString()
  createdAt!: string;

  @IsDateString()
  updatedAt!: string;

  @IsNumber()
  commentsCount!: number;
}
