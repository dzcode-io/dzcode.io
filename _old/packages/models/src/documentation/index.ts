import { Type } from "class-transformer";
import { IsString, ValidateNested } from "class-validator";
import { BaseEntity, Model } from "src/_base";
import { AccountEntity } from "src/account";

export class DocumentationInfoEntity extends BaseEntity {
  @IsString()
  slug!: string;

  @IsString()
  title!: string;
}

export class DocumentationEntity extends DocumentationInfoEntity {
  @IsString()
  image!: string;

  @IsString()
  description!: string;

  @IsString()
  content!: string;

  @ValidateNested({ each: true })
  @Type(() => AccountEntity)
  authors!: Model<AccountEntity>[];

  @ValidateNested({ each: true })
  @Type(() => AccountEntity)
  contributors!: Model<AccountEntity>[];
}
