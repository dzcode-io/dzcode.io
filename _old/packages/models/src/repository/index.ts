import { Type } from "class-transformer";
import { IsIn, IsNumber, IsString, ValidateNested } from "class-validator";
import { BaseEntity, Model } from "src/_base";
import { AccountEntity } from "src/account";
import { ContributionEntity } from "src/contribution";

export class RepositoryStatsEntity extends BaseEntity {
  @IsNumber()
  contributionCount!: number;

  @IsString({ each: true })
  languages!: string[];
}

const RepositoryProviders = ["github", "gitlab"] as const;
type RepositoryProvider = typeof RepositoryProviders[number];

export class RepositoryEntity extends BaseEntity {
  @IsIn(RepositoryProviders)
  provider!: RepositoryProvider;

  @IsString()
  owner!: string;

  @IsString()
  repository!: string;

  @Type(() => RepositoryStatsEntity)
  stats?: Model<RepositoryStatsEntity>;

  @ValidateNested({ each: true })
  @Type(() => AccountEntity)
  contributors?: Model<AccountEntity>[];

  @ValidateNested({ each: true })
  @Type(() => ContributionEntity)
  contributions?: Model<ContributionEntity>[];
}
