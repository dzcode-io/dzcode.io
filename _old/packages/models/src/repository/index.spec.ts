import { runDTOTestCases } from "src/_test";

import { RepositoryEntity } from ".";

runDTOTestCases(
  RepositoryEntity,
  {
    provider: "github",
    owner: "dzcode-io",
    repository: "leblad",
  },
  {
    contributions: [],
    contributors: [],
    stats: { contributionCount: 10, languages: ["TypeScript", "Rust"] },
  },
);
