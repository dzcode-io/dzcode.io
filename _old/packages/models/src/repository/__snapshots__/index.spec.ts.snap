// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`should match snapshot when providing all fields: errors 1`] = `[]`;

exports[`should match snapshot when providing all fields: output 1`] = `
RepositoryEntity {
  "contributions": [],
  "contributors": [],
  "owner": "dzcode-io",
  "provider": "github",
  "repository": "leblad",
  "stats": RepositoryStatsEntity {
    "contributionCount": 10,
    "languages": [
      "TypeScript",
      "Rust",
    ],
  },
}
`;

exports[`should match snapshot when providing required fields only: errors 1`] = `[]`;

exports[`should match snapshot when providing required fields only: output 1`] = `
RepositoryEntity {
  "owner": "dzcode-io",
  "provider": "github",
  "repository": "leblad",
}
`;

exports[`should show an error that matches snapshot when passing empty object: errors 1`] = `
[
  ValidationError {
    "children": [],
    "constraints": {
      "isIn": "provider must be one of the following values: github, gitlab",
    },
    "property": "provider",
    "target": RepositoryEntity {},
    "value": undefined,
  },
  ValidationError {
    "children": [],
    "constraints": {
      "isString": "owner must be a string",
    },
    "property": "owner",
    "target": RepositoryEntity {},
    "value": undefined,
  },
  ValidationError {
    "children": [],
    "constraints": {
      "isString": "repository must be a string",
    },
    "property": "repository",
    "target": RepositoryEntity {},
    "value": undefined,
  },
]
`;

exports[`should show an error that matches snapshot when passing empty object: output 1`] = `RepositoryEntity {}`;
